toolbox/oncall_alert/main.go

```diff
@@ -172,91 +175,108 @@ func formatMessage(failures map[*postSubmitJob]bool) (mess string) {
-func getLatestRun(job string) (int, error) {
-	lastBuildFile := filepath.Join(job, lastBuildTXT)
-	latestBuildString, err := gcsClient.GetFileFromGCSString(*gcsBucket, lastBuildFile)
+func getLatestRun(jobName string) (int, error) {
+	lastBuildFile := filepath.Join(jobName, lastBuildTXT)
+	latestBuildString, err := gcsClient.Read(*gcsBucket, lastBuildFile)
 	if err != nil {
 		return 0, err
 	}
 	latestBuildInt, err := strconv.Atoi(latestBuildString)
 	if err != nil {
-		log.Printf("Failed to convert %s to int: %v", latestBuildString, err)
+		log.Printf("Failed to convert %s to int: %v\n", latestBuildString, err)
 		return 0, err
 	}
 	return latestBuildInt, nil
 }
 
-// Check the latest run of each protected Post-Submit tests.
+// From last check onward, check all the runs of protected Post-Submit tests.
 // Record failure if it hasn't been recorded and update latestRun if necessary
-func updatePostSubmitStatus() (map[*postSubmitJob]bool, int) {
-	// If a "new" failure appears in this circle, put it in and with value true
-	// So if no new failure, this map should be empty --> no new notification
-	newFailures := make(map[*postSubmitJob]bool)
-
-	// Total count of jobs which is in failing status right now.
-	// Including jobs failed in circles before.
-	// If 0 --> unblock auto-merge
-	failedCount := 0
-
-	for _, job := range postSubmitJobs {
-		latestRunNo, err := getLatestRun(job.name)
+// Returns an array of postSubmitJob since the same job might fail at multiple runs
+func checkOnJobsWatched() []failure {
+	newFailures := []failure{}
+	for _, job := range jobsWatched {
+		CurrentRunNo, err := getLatestRun(job.name)
 		if err != nil {
-			log.Printf("Failed to get last run number of %s: %v", job.name, err)
+			log.Printf("Failed to get last run number of %s: %v\n", job.name, err)
 			continue
 		}
-
-		if *debug {
-			log.Printf("Job: %s -- Latest Run No: %d -- Recorded Run No: %d", job.name, latestRunNo, job.latestRun)
+		log.Printf("Job: [%s] \t Current Run No: [%d] \t Previously Checked: [%d]\n",
+			job.name, CurrentRunNo, job.lastCheckedRunNo)
+		// Avoid pulling entire history when the daemon has just started
+		if job.lastCheckedRunNo == 0 {
+			job.lastCheckedRunNo = CurrentRunNo - 1
 		}
-		// New test finished for this job in this circle
-		if latestRunNo > job.latestRun {
-			prowResult, err := getProwResult(filepath.Join(job.name, strconv.Itoa(latestRunNo)))
-			if err != nil {
-				log.Printf("Failed to get prowResult %s/%d", job.name, latestRunNo)
-				continue
-			}
-			job.latestRun = latestRunNo
-			job.latestRunPass = prowResult.Passed
-			if *debug {
-				log.Printf("Job: %s -- Latest Run No: %d -- Passed? %t", job.name, latestRunNo, prowResult.Passed)
+		// Clone a slice of keys in map to avoid editing while iterating
+		// Then check previously pending runs to see if results are ready
+		pendingRuns := make([]int, 0, len(job.pendingFirstRuns))
+		for runNo := range job.pendingFirstRuns {
```

Comment:
you meant to use an array ?
```golang
var pendingRuns [10]int
	for i:=0; i<10; i++ {
		pendingRuns[i] = i
	}
```golang

Reply:
I tried your suggestion but no luck. The reason is that `pendingFirstRuns` has type `map[int]time.Time`. The size of the key set is unknown at compile time so I cannot do things like `var pendingRuns [10]int`. Using make to claim the size that fits exactly our need and then calling append() will not resize the array.